index.ts

```diff
@@ -152,186 +213,210 @@ module.exports = {};`,
+      dep.baseDir = path.join(this.baseDir, 'node_modules', dep.name);
+      dep.writeSync();
     }
   }
 
-  writeSync(root = this.root) {
-    fixturify.writeSync(root, this.toJSON());
+  static fromDir(root: string, opts?: ReadDirOpts): Project {
+    let project = new Project();
+    project.readSync(root, opts);
+    return project;
   }
 
-  readSync(root = this.root) {
-    const files = unwrapPackageName(fixturify.readSync(root), this.name);
+  private readSync(root: string, opts?: ReadDirOpts): void {
+    const files = fixturify.readSync(root, {
+      // when linking deps, we don't need to crawl all of node_modules
+      ignore: opts?.linkDeps ? ['node_modules'] : [],
+    });
 
     this.pkg = deserializePackageJson(getFile(files, 'package.json'));
-    const nodeModules = getFolder(files, 'node_modules');
-
-    // drop "special files"
-    delete files['node_modules'];
     delete files['package.json'];
-
-    this._dependencies = {};
-    this._devDependencies = {};
     this.files = files;
 
-    keys(this.pkg.dependencies).forEach(dependency => {
-      this.addDependency((this.constructor as ProjectConstructor).fromJSON(nodeModules, dependency));
-    });
+    if (opts?.linkDeps) {
+      if (this.pkg.dependencies) {
+        for (let dep of Object.keys(this.pkg.dependencies)) {
+          this.linkDependency(dep, { baseDir: path.join(root, this.name) });
+        }
+      }
+      if (this.pkg.devDependencies) {
+        for (let dep of Object.keys(this.pkg.devDependencies)) {
+          this.linkDevDependency(dep, { baseDir: path.join(root, this.name) });
+        }
+      }
+    } else {
+      const nodeModules = getFolder(files, 'node_modules');
+      delete files['node_modules'];
+      keys(this.pkg.dependencies).forEach(dependency => {
+        this.addDependency(
+          new (this.constructor as typeof Project)({ files: unwrapPackageName(nodeModules, dependency) })
+        );
+      });
+      keys(this.pkg.devDependencies).forEach(dependency => {
+        this.addDevDependency(
+          new (this.constructor as typeof Project)({ files: unwrapPackageName(nodeModules, dependency) })
+        );
+      });
+    }
+  }
 
-    keys(this.pkg.devDependencies).forEach(dependency => {
-      this.addDevDependency((this.constructor as ProjectConstructor).fromJSON(nodeModules, dependency));
-    });
+  addDependency(name?: string, version?: string, args?: Omit<ProjectArgs, 'name' | 'version'>): Project;
+  addDependency(name?: string, args?: Omit<ProjectArgs, 'name'>): Project;
+  addDependency(args?: ProjectArgs): Project;
+  addDependency(args?: Project): Project;
+  addDependency(
+    first?: string | ProjectArgs | Project,
+    second?: string | Omit<ProjectArgs, 'name'>,
+    third?: Omit<ProjectArgs, 'name' | 'version'>
+  ): Project {
+    return this.addDep(first, second, third, '_dependencies');
   }
 
-  addDependency(name: string | Project, version?: string, cb?: (project: Project) => void) {
```

Comment:
Similarly here, `cb?: (project: Project) => void)` is an often used API, making this breaking change makes upgrading much more tedious than it needs to be.  Could you help me understand rationale?